xla/service/gpu/transforms/gemm_rewriter.cc

```diff
@@ -361,28 +362,67 @@ std::optional<MatchedFp8Param> MatchFp8Param(HloInstruction *instr) {
 // dimension. There must be only one contracting and only one non-contracting
 // dimension. Keeps the layout the same.
 HloInstruction *TransposeMatrix(HloInstruction *instr, int64_t contracting_dim,
-                                absl::Span<const int64_t> batch_dims) {
+                                absl::Span<const int64_t> batch_dims,
+                                bool col_maj = false) {
   // Identify the dimensional order which describes a transpose of the
   // contracting and non-contracting dimensions of the GEMM.
   std::vector<int64_t> permutation(instr->shape().dimensions_size(), -1);
-  // Discard the batch dimensions.
```

Comment:
I believe if you move this down below the "Identify the non-contracting dimension" block, it will break the logic.
See the check

```
 if (permutation[i] == -1 && contracting_dim != i) {
      non_contracting_dim = i;
    }
```

We rely on the permutation vector to contain values >= 0 for the batch dimensions, otherwise we might detect the first batch dimension as non_contracting_dim.

Reply:
You're right, that is my overlook. Fixed.